File: data-structure/LinkedList/LinkedList.ts
import { LinkedListNode } from './ILinkedList';

export default class LinkedList<E> {
  public dummyHead: LinkedListNode<E> = new LinkedListNode<E>();
  private size: number = 0;

  /**
   * 判断链表是否为空
   * Time Complexity O(1)
   * Space Complexity O(1)
   * @return {boolean}
   */
  isEmpty(): boolean {
    return this.size === 0;
  }

  /**
   * 获取当前链表中元素的个数
   * Time Complexity O(1)
   * Space Complexity O(1)
   * @return {boolean}
   */
  getSize(): number {
    return this.size;
  }

  /**
   * 获取当前链表中index位置添加新的元素e
   * 这里需要找到index位置的前一个元素，让前一个元素的next指向新的元素，新的元素的next指向prev.next
   * Time Complexity O(n)
   * Space Complexity O(1)
   * @param {number} index 元素插入的位置
   * @param {E} e 要插入的元素
   * @return {void}
   */
  add(index: number, e: E): void {
    if (index < 0 || index > this.size) {
      throw new Error('Add failed. Illegal index.');
    }

    let prev = this.dummyHead;
    for (let i = 0; i < index; i++) {
      prev = prev.next;
    }

    prev.next = new LinkedListNode<E>(e, prev.next);
    this.size++;
  }

  /**
   * 在链表头添加一个元素
   * Time Complexity O(1)
   * Space Complexity O(1)
   * @param {E} e 要插入链表的元素
   * @return {void}
   */
  addFirst(e: E): void {
    this.add(0, e);
  }

  /**
   * 在链表尾部添加一个元素
   * Time Complexity O(n)
   * Space Complexity O(1)
   * @param {E} e 要插入链表的元素
   * @return {void}
   */
  addLast(e: E): void {
    this.add(this.size, e);
  }

  /**
   * 获取链表的第index个位置的元素
   * Time Complexity O(n)
   * Space Complexity O(1)
   * @param {number} index 要获取元素的索引
   * @return {E}
   */
  get(index: number): E {
    if (index < 0 || index > this.size) {
      throw new Error('Get failed. Illegal index.');
    }

    let cur = this.dummyHead.next;
    for (let i = 0; i < index; i++) {
      cur = cur.next;
    }
    return cur.e;
  }

  /**
   * 获取链表头的元素
   * Time Complexity O(1)
   * Space Complexity O(1)
   * @return {E}
   */
  getFirst(): E {
    return this.get(0);
  }

  /**
   * 获取链表尾的元素
   * Time Complexity O(1)
   * Space Complexity O(1)
   * @return {E}
   */
  getLast(): E {
    return this.get(this.size - 1);
  }

  /**
   * 修改链表第index位置的元素为e
   * Time Complexity O(n)
   * Space Complexity O(1)
   * @param {number} index 要修改的元素的索引
   * @param e 要被修改成的元素
   * @return {void}
   */
  set(index: number, e: E): void {
    if (index < 0 || index >= this.size) {
      throw new Error('Set failed. Illegal index.');
    }

    let cur = this.dummyHead.next;
    for (let i = 0; i < index; i++) {
      cur = cur.next;
    }
    cur.e = e;
  }

  /**
   * 查找链表中是否包含元素e
   * Time Complexity O(n)
   * Space Complexity O(1)
   * @param e 需要查找的元素
   * @return {boolean}
   */
  contains(e: E): boolean {
    let cur = this.dummyHead.next;
    while (cur !== null) {
      if (cur.e === e) {
        return true;
      }
      cur = cur.next;
    }
    return false;
  }

  /**
   * 删除链表中第index位置的索引，并返回删除的元素
   * Time Complexity O(n)
   * Space Complexity O(1)
   * @param {number} index 需要删除的元素所在的位置
   * @return {E}
   */
  remove(index: number): E {
    if (index < 0 || index >= this.size) {
      throw new Error('Remove failed. Index is illegal.');
    }
    let prev = this.dummyHead;
    for (let i = 0; i < index; i++) {
      prev = prev.next;
    }
    let delNode = prev.next;
    prev.next = delNode.next;
    delNode.next = null; // 断开delNode与其下一个节点的连接
    this.size--;
    return delNode.e;
  }

  /**
   * 删除链表头的元素
   * Time Complexity O(1)
   * Space Complexity O(1)
   * @return {E}
   */
  removeFirst(): E {
    return this.remove(0);
  }

  /**
   * 删除链表尾的元素
   * Time Complexity O(n)
   * Space Complexity O(1)
   * @return {E}
   */
  removeLast(): E {
    return this.remove(this.size - 1);
  }

  /**
   * 删除链表中的元素e
   * Time Complexity O(n)
   * Space Complexity O(1)
   * @param {E} e 要删除的元素e
   * @return {E}
   */
  removeElement(e: E): void {
    let prev: LinkedListNode<E> = this.dummyHead;
    while (prev.next !== null) {
      if (prev.next.e === e) {
        break;
      }
      prev = prev.next;
    }

    if (prev.next !== null) {
      let delNode = prev.next;
      prev.next = delNode.next;
      delNode.next = null;
      this.size--;
    }
  }
}

File: DataStructureJava/LinkedList.java
/**
 * 链表的特点
 * 1. 最简单的动态数据结构
 * 2. 更深入的理解引用（指针）
 * 3. 更深入的理解递归
 * 4. 辅助组成其它数据结构
 * 5. 数据存储在节点(node)中
 *
 * 1 -> 2 -> 3 -> NULL
 *
 * 优点：真正的动态，不需要处理固定容量的问题
 * 缺点：丧失了随机访问的能力
 */
public class LinkedList<E> {
    private class Node {
        public E e;
        public Node next;

        public Node(E e, Node next) {
            this.e = e;
            this.next = next;
        }

        public Node(E e) {
            this(e, null);
        }

        public Node() {
            this(null, null);
        }

        @Override
        public String toString() {
            return e.toString();
        }
    }

    private Node dummyHead;
    private int size;

    public LinkedList() {
        // 这里初始化是Node传入空值即可
        // 因为初始时链表中没有元素，当添加元素时，才会给dummyHead指向下一个元素
        dummyHead = new Node();
        size = 0;
    }

    public int getSize() {
        return size;
    }

    public boolean isEmpty() {
        return size == 0;
    }

    // 在链表头添加元素
    public void addFirst(E e) {
        add(0, e);
    }

    // 在链表的index位置添加新的元素e
    public void add(int index, E e) {
        if (index < 0 || index > size)
            throw new IllegalArgumentException("Add Failed. Illegal index");
        
        Node prev = dummyHead;
        for (int i = 0; i < index; i++) {
            prev = prev.next;
        }
        prev.next = new Node(e, prev.next);
        size++;
    }

    // 在链表末尾添加元素e
    public void addLast(E e) {
        add(size, e);
    }

    // 获取链表中第index位置的元素
    public E get(int index) {
        if (index < 0 || index >= size)
            throw new IllegalArgumentException("Get failed. Illegal index");
        Node cur = dummyHead.next;
        for (int i = 0; i < index; i++) {
            cur = cur.next;
        }
        return cur.e;
    }

    // 获取链表的第一个元素
    public E getFirst() {
        return get(0);
    }

    // 获取链表的最后一个元素
    public E getLast() {
        return get(size - 1);
    }

    // 修改链表的第index位置的元素
    public void set(int index, E e) {
        if (index < 0 || index >= size)
            throw new IllegalArgumentException("Set failed. Illegal index");
        Node cur = dummyHead.next;
        for (int i = 0; i < index; i++)
            cur = cur.next;
        cur.e = e;
    }

    // 查找链表中是否包含元素e
    public boolean contains(E e) {
        Node cur = dummyHead.next;
        while (cur != null) {
            if (cur.e == e) {
                return true;
            }
            cur = cur.next;
        }
        return false;
    }

    // 删除链表中第index位置的元素
    public E remove(int index) {
        if (index < 0 || index >= size)
            throw new IllegalArgumentException("Remove failed. Illegal index");
        Node prev = dummyHead;
        for (int i = 0; i < index; i++)
            prev = prev.next;
        Node retNode = prev.next;
        prev.next = retNode.next;
        retNode.next = null;
        size--;
        return retNode.e;
    }

    // 删除链表的第一个元素
    public E removeFirst() {
        return remove(0);
    }

    // 删除链表的最后一个元素
    public E removeLast() {
        return remove(size - 1);
    }

    // 从链表中删除元素e
    public void removeElement(E e) {
        Node prev =dummyHead;
        while (prev.next != null) {
            if (prev.next.e.equals(e))
                break;
            prev = prev.next;
        }
        if (prev.next != null) {
            Node delNode = prev.next;
            prev.next = delNode.next;
            delNode.next = null;
            size--;
        }
    }

    @Override
    public String toString() {
        StringBuilder res = new StringBuilder();
        Node cur = dummyHead.next;
        while (cur != null) {
            res.append(cur + "->");
            cur = cur.next;
        }
        res.append("NULL");
        return res.toString();
    }

    public static void main(String[] args) {
        LinkedList<Integer> linkedList = new LinkedList<>();
        for (int i = 0; i < 5; i++) {
            linkedList.addFirst(i);
            System.out.println(linkedList);
        }

        linkedList.add(2, 666);
        System.out.println(linkedList);

        linkedList.remove(2);
        System.out.println(linkedList);

        linkedList.removeFirst();
        System.out.println(linkedList);

        linkedList.removeLast();
        System.out.println(linkedList);
    }
}
